filechooser-portal/Main.vala

```diff
@@ -232,9 +218,20 @@ public class Files.FileChooserPortal : Object {
                     if (dialog.filter != null) {
                         _results["current_filter"] = dialog.filter.to_gvariant ();
                     }
-
-                    _response = 0;
-                    break;
+                    var chosen_uri = dialog.get_uri ();
+                    if (chosen_uri != supplied_uri) {
+                        check_overwrite_uri.begin (chosen_uri, dialog, (obj, res) => {
+                            if (check_overwrite_uri.end (res)) {
+                                _response = 0;
+                                save_file.callback ();
+                            }
+                        });
+
+                        return; // Continue showing dialog until check completes
+                    } else { // No need to check full uri supplied by calling app
+                        _response = 0;
+                        break;
+                    }
```

Comment:
I think breaking early here make the code more readable: (suggestion already have the changes needed by the `check_overwrite_file/create_overwrite_dialog` sugesstion) 
```suggestion
                   _response = 0;

                    var chosen_file = dialog.get_file ();
                    if (!file.query_exists () || chosen_file.get_uri () == supplied_uri) {
                        break; // No need to check full uri supplied by calling app
                    }
                    
                    var overwrite_dialog = create_overwrite_dialog (dialog, chosen_file);
                    overwrite_dialog.response.connect ((response) => {
                        if (response == Gtk.ResponseType.YES) {
                            save_file.callback ();
                        } else {
                            _results.remove_all ();
                            _response = 2;
                        }
                        
                        overwrite_dialog.destroy ();
                    });
                    overwrite_dialog.present ();
                    return; // Continue showing dialog until check completes
```